lib/coverband/adapters/hash_redis_store.rb

```diff
@@ -0,0 +1,164 @@
+        roots = Coverband.configuration.all_root_paths
+        absolute_path = relative_path_to_full(file, roots)
+        file_hash = file_hash(absolute_path)
+        relative_path_file = full_path_to_relative(file)
+        Coverband::TYPES.each do |type|
+          @redis.del(key(relative_path_file, type, file_hash: file_hash))
+        end
+        @redis.srem(files_key, relative_path_file)
+      end
+
+      def save_report(report)
+        roots = Coverband.configuration.all_root_paths
+        report_time = Time.now.to_i
+        updated_time = type == Coverband::EAGER_TYPE ? nil : report_time
+        script_id = hash_incr_script
+        @redis.pipelined do
+          keys = report.map do |file, data|
+            relative_file = full_path_to_relative(file)
+            absolute_path = relative_path_to_full(file, roots)
+            file_hash = file_hash(absolute_path)
+            key = key(relative_file, file_hash: file_hash)
+            script_input = save_report_script_input(
+              key: key,
+              file: relative_file,
+              file_hash: file_hash,
+              data: data,
+              report_time: report_time,
+              updated_time: updated_time
+            )
+            @redis.evalsha(script_id, script_input[:keys], script_input[:args])
+            key
+          end
+          @redis.sadd(files_key, keys) if keys.any?
+        end
+      end
+
+      def coverage(local_type = nil)
+        roots = Coverband.configuration.all_root_paths
+        keys = files_set(local_type)
+        keys.each_with_object({}) do |key, hash|
+          data_from_redis = @redis.hgetall(key)
+
+          next if data_from_redis.empty?
+
+          file = data_from_redis[FILE_KEY]
+          stored_hash = data_from_redis[FILE_HASH]
+          absolute_path = relative_path_to_full(file, roots)
+          next unless file_hash(absolute_path) == stored_hash
+
+          max = data_from_redis[FILE_LENGTH_KEY].to_i - 1
+          data = Array.new(max + 1) do |index|
+            line_coverage = data_from_redis[index.to_s]
+            line_coverage.nil? ? nil : line_coverage.to_i
+          end
+          hash[file] = data_from_redis.select { |meta_data_key, _value| META_DATA_KEYS.include?(meta_data_key) }.merge!('data' => data)
+          hash[file][LAST_UPDATED_KEY] = hash[file][LAST_UPDATED_KEY].to_i
+          hash[file][FIRST_UPDATED_KEY] = hash[file][FIRST_UPDATED_KEY].to_i
+        end
+      end
+
+      private
+
+      def save_report_script_input(key:, file:, file_hash:, data:, report_time:, updated_time:)
+        data.each_with_index
+            .each_with_object(keys: [key], args: [report_time, updated_time, file, file_hash, @ttl, data.length]) do |(coverage, index), hash|
+          if coverage
+            hash[:keys] << index
+            hash[:args] << coverage
+          end
+        end
+      end
+
+      def hash_incr_script
+        @hash_incr_script ||= @redis.script(:load, <<~LUA)
```

Comment:
perhaps load this from a file or something to make it easier to read?

Reply:
I could do that but I have string interpolation here. Think it's worth it?

Reply:
Nah, we can leave as is for now